Source/DafnyCore/Backends/ExecutableBackend.cs

```diff
@@ -50,17 +72,27 @@ public abstract class ExecutableBackend : IExecutableBackend {
   }
 
   protected void ProcessOuterModules(Program dafnyProgram) {
-    var outerModules = GetOuterModules();
-    ModuleDefinition rootUserModule = null;
-    foreach (var outerModule in outerModules) {
-      var newRoot = new ModuleDefinition(RangeToken.NoToken, new Name(outerModule), new List<IToken>(),
-        ModuleKindEnum.Concrete, false,
-        null, null, null);
-      newRoot.EnclosingModule = rootUserModule;
-      rootUserModule = newRoot;
+    // Apply the --outer-module option from any translation records for libraries,
+    // but only to top-level modules.
+    var outerModules = new Dictionary<string, ModuleDefinition>();
+    foreach (var module in dafnyProgram.CompileModules) {
+      if (module.EnclosingModule is not DefaultModuleDefinition) {
+        continue;
+      }
+
+      var recordedOuterModuleName = (string)dafnyProgram.Compilation.AlreadyTranslatedRecord.Get(null, module.FullDafnyName, OuterModule);
```

Comment:
At some point we should use the translation records to configure the options of the library modules, so consuming code does not have to look at `Compilation.AlreadyTranslatedRecord` in addition to `DafnyOptions`. In this case this would lead to:

```
var outerModule = module.Options.Get(OuterModule);
var outerModule = outerModules.GetOrCreate(recordedOuterModuleName, () => CreateOuterModule(recordedOuterModuleName));
module.EnclosingModule = outerModule;
```

With a change in `ProgramParser` so that the option `OuterModule` is only set for the default module.

Note that currently module options are only stored in `ModuleDecl`, not in `ModuleDefinition`, but ProgramParser could also store them in ModuleDefinition.

Reply:
I was thinking that too, in fact that's part of the reason I tried to add the processing to the parsing-related code in the first pass. :)